### PR TITLE
Dm 4484 marker tool - create drawing layer for marker tool

### PR DESCRIPTION
**Pull request by cwang2016**
The implementation includes:
- DrawObj for marker
- Marker dropdown list under the marker icon to show the marker and footprint items (will be added later). 
- Marker drawing layer rendering and operation including action creator and action dispatch functions.
- Marker UI component shown in Drawing Layers popup. 
The implementation set up some work which can be similarly expanded to footprints in the future. 
 
Maker drawing layer operation:
- select 'Add Marker' from dropdown list to add a  new drawing layer
- click anywhere to locate the newly added marker
- click and drag the mouse  to relocate or resize the marker: 
  for relocation: click and drag inside the circle, then drag the mouse.
  for resize: click inside the circle, then click and drag the handler to resize 
- label and its location are set from Marker tool UI

### CODE REVIEW

File: src/firefly/js/visualize/DrawLayerCntlr.js [hunk 1]
```diff
@@ -50,6 +54,12 @@ const REGION_DELETE_LAYER = `${DRAWLAYER_PREFIX}.RegionPlot.deleteLayer`;
 const REGION_ADD_ENTRY = `${DRAWLAYER_PREFIX}.RegionPlot.addRegion`;
 const REGION_REMOVE_ENTRY = `${DRAWLAYER_PREFIX}.RegionPlot.removeRegion`;
 
+// makler and footprint
```
Comment (robyww) [inline src/firefly/js/visualize/DrawLayerCntlr.js]: picky: marker not makler

File: src/firefly/js/visualize/DrawLayerCntlr.js [hunk 2]
```diff
@@ -50,6 +54,12 @@ const REGION_DELETE_LAYER = `${DRAWLAYER_PREFIX}.RegionPlot.deleteLayer`;
...
+const MARKER_START = `${DRAWLAYER_PREFIX}.MarkerTool.markerStart`;
+const MARKER_MOVE = `${DRAWLAYER_PREFIX}.MarkerTool.markerMove`;
+const MARKER_END = `${DRAWLAYER_PREFIX}.MarkerTool.markerEnd`;
+const MARKER_CREATE= `${DRAWLAYER_PREFIX}.MarkerTool.markerCreate`;
+
 export const DRAWING_LAYER_KEY= 'drawLayers';
 
 const clone = (obj,params={}) => Object.assign({},obj,params);
```
Comment (robyww) [inline src/firefly/js/visualize/DrawLayerCntlr.js]: that is now a clone in `WebUtil.js`

**Comment by robyww** [comment]
UI Comments: 

 - Mostly it looks the like the old one and works better.
 - On the Layer control: `show: <title>`, It the title should change as you enter text for the label. For example, Marker-1 would be how it started but when the  user enters 'my favorite' it should change to that.
 - The mouse should change to a pointer when you are on a marker.  If you want to be really fancy it could change to a resize cursor on each resize box.
 - The size is not in world coordinates, the position but the size is not.  When I zoom the image, the circle size does not change with the zoom level.

**Comment by cwang2016** [comment]
more improvement,
- on the layer control, show: <title> and the title change as you enter the text for the label. 
- the mouse changes to a pointer once the cursor moves onto the marker, and the mouse becomes a resize sign when the cursor moves on the handler of the marker. 
- When the cursor becomes a pointer, the marker can be relocated by dragging the mouse,  and when the cursor becomes a resize one,  the marker can be resized by dragging the mouse. 
- the marker size changes as the image is zoomed.

**Comment by ejoliet** [comment]
Nice improvement!

On 6/16/16 11:39 PM, Cindy Wang wrote:
>
> more improvement,
>
>   * on the layer control, show: and the title change as you enter the
>     text for the label. </li> <li>the mouse changes to a pointer once
>     the cursor moves onto the marker, and the mouse becomes a resize
>     sign when the cursor moves on the handler of the marker. </li>
>     <li>When the cursor becomes a pointer, the marker can be relocated
>     by dragging the mouse, and when the cursor becomes a resize one,
>     the marker can be resized by dragging the mouse. </li> <li>the
>     marker size changes as the image is zoomed. </li> </ul>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Caltech-IPAC/firefly/pull/104#issuecomment-226692187>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AAvATK40ImgeVNvB3DPa9o5kujKfBsL8ks5qMkEogaJpZM4I3DDD>.
>

-- 
Emmanuel Joliet
NASA/IPAC  Infrared Science Archive
MS 100-22, Caltech, Pasadena, CA 91125
Office: 170
Email:  ejoliet@ipac.caltech.edu
Phone:  (626) 395-1489

**Pull request merged**